### PR TITLE
Add validation of Docker Images

### PR DESCRIPTION
**Pull request by AndreyKoltsov1997**
The pull request introduces:
- Post-Push validation of Docker images. As for now, check for image's size regression had been implemented.
- Automation utilities (Kotlin script) implemented for validation purposes.
- Integration of validation into build config-generation logic written on C#.
- Documentation update for existing code aimed at generation of build configuration. 

In order to simplify review process, the PR does not include re-generated TeamCity build configurations. In my opinion, we could add it once the changes are merged or reviewed.

### CODE REVIEW

**Comment by AndreyKoltsov1997** [comment]
Closing the pull request until completion in order to prevent redundant e-mail notifications for the subscribers to the repository. Please, excuse me for such excessive amount of notifications recently.
Will be re-opened once validation part is complete and tested.

File: tool/automation/README.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    <docker image fully-qualified domain name> \
+    <(optional) previous docker image fully-qualified domain name>
+```
+
+### 2.1 TeamCity statistics
+Image size is reported into TeamCity via Servce Message using the following parretn: `"SIZE-$currentName"`.
+
+
+### 2.2 A note on automation's dependencies / assumptions
+Automtion tools has the following assumptions that might be changed over the time / releases:
```
Comment (burnasheva) [inline tool/automation/README.md:21]: typo

File: tool/automation/README.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+## 1. Purpose
+Utilities that simplify automation of releasing TeamCity docker images.
+
+
+## 2. Image Validation
+[ImageValidation.kts](ImageValidation.kts) - responsible for validation of Docker images.
+**Usage**:
+```
+kotlic -script ImageValidation.kts \
```
Comment (burnasheva) [inline tool/automation/README.md:11]: typo

File: tool/README.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+# 1. Dockerfiles & Build Configurations generator
+[TeamCity.Docker](TeamCity.Docker) is an utility application (C#) that generates Dockerfiles along with build configurations for
+build up and publishing of TeamCity Docker images for different platforms.
+
+## 1.1 Configuration options
+
+The program accepts verbs (main options), as well as their configuration (sub-options).
+
+`generate` Generate docker and readme files.
```
Comment (burnasheva) [inline tool/README.md:14]: missed '-'

File: tool/README.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+| Option | short flag | Required | Description |
+| --- | --- | --- | ------ |
+| `source` | `-s` | false | Path to configuration directory. |
+| `files` | `-f` | false | Semicolon separated configuration file. Separator - `;` |
+| `context` | `-c` | false | Path to the context directory. |
+| `docker` | `-d` | false | The Docker Engine endpoint like tcp://localhost:2375 (defaults: npipe://./pipe/docker_engine fo windows and unix:///var/run/docker.sock for others). |
+| `verbose` | `-v` | false | Specified neccessity for verbose (detailed) output. |
+| `regex` | `-r` | false | Regular expression for filtering an internal build graph. Used for development purposes mostly. |
+
+# 2. Automation utilities
```
Comment (burnasheva) [inline tool/README.md:50]: typo

File: tool/README.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+[Automation utiltiies](automation) contains utilities that simplify automation of releasing TeamCity docker images.
```
Comment (burnasheva) [inline tool/README.md:52]: typo

Comment (burnasheva) [reply tool/README.md:52]: I'd suggest using "simplify release process" instead of "simplify automation of releasing".

File: tool/README.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
+# TeamCity Docker Images - tools
+Tools that are used within TeamCity Docker images release process.
+
+
+
+# 1. Dockerfiles & Build Configurations generator
+[TeamCity.Docker](TeamCity.Docker) is an utility application (C#) that generates Dockerfiles along with build configurations for
```
Comment (burnasheva) [inline tool/README.md:7]: "along with build configuration" -> "along with Kotlin DSL for build configurations" sounds better, what do you think?

File: tool/TeamCity.Docker/GenerateCommand.cs [hunk 7]
```diff
@@ -46,6 +50,7 @@ public Task<Result> Run()
                 return Task.FromResult(Result.Error);
             }
 
+            // conver configuration parameters into grapth for further processing
```
Comment (burnasheva) [inline tool/TeamCity.Docker/GenerateCommand.cs:53]: typo?

File: tool/TeamCity.Docker/TeamCityKotlinSettingsGenerator.cs [hunk 8]
```diff
@@ -39,6 +40,15 @@ internal class TeamCityKotlinSettingsGenerator : IGenerator
             _nodesDescriptionsFactory = nodesDescriptionFactory ?? throw new ArgumentNullException(nameof(nodesDescriptionFactory));
         }
 
+        /// <summary>
+        /// Generates the following TeamCity Build Configurations (KotlinDSL) responsible for ...
```
Comment (burnasheva) [inline tool/TeamCity.Docker/TeamCityKotlinSettingsGenerator.cs:44]: Could you please re-check the summary (it looks like it has duplicates in it)?

File: tool/TeamCity.Docker/TeamCityKotlinSettingsGenerator.cs [hunk 9]
```diff
@@ -301,6 +340,68 @@ private IEnumerable<string> CreatePushBuildConfiguration(string buildTypeId, str
             yield return string.Empty;
         }
 
+        /// <summary>
+        /// Generates Kotlin DSL file with build configuration for post-push Docker image check.
+        /// A post-push validation build had been done in purpose of lower cost for failure within build chain.
```
Comment (burnasheva) [inline tool/TeamCity.Docker/TeamCityKotlinSettingsGenerator.cs:345]: .. for the purpose of...

File: tool/TeamCity.Docker/TeamCityKotlinSettingsGenerator.cs [hunk 10]
```diff
@@ -301,6 +340,68 @@ private IEnumerable<string> CreatePushBuildConfiguration(string buildTypeId, str
...
+        /// It includes checks needed for service purposes only.
```
Comment (burnasheva) [inline tool/TeamCity.Docker/TeamCityKotlinSettingsGenerator.cs:346]: Could you please clarify which service we mean here?

File: tool/TeamCity.Docker/TeamCityKotlinSettingsGenerator.cs [hunk 11]
```diff
@@ -301,6 +340,68 @@ private IEnumerable<string> CreatePushBuildConfiguration(string buildTypeId, str
...
+                yield return dependencies;
+            }
+
+
+            yield return "})";
+            yield return string.Empty;
+        }
+
+        /// <summary>
+        /// Generates TeamCity build configuration (Kotlin DSL) for publishment of Docker image manifests.
```
Comment (burnasheva) [inline tool/TeamCity.Docker/TeamCityKotlinSettingsGenerator.cs:396]: publishing

File: tool/TeamCity.Docker/TeamCityKotlinSettingsGenerator.cs [hunk 12]
```diff
@@ -666,6 +830,12 @@ private IEnumerable<string> CreatePushCommand(string imageId, string name, param
             yield return string.Empty;
         }
 
+        /// <summary>
+        /// Constructs Kotlin DSL's dockerComamnd {...} for image re-tag.
+        /// </summary>
+        /// <param name="repoTag">original Docker iamge tag</param>
```
Comment (burnasheva) [inline tool/TeamCity.Docker/TeamCityKotlinSettingsGenerator.cs:836]: typo

File: tool/automation/ImageValidation.kts [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+
+/**
+ * Validates Docker image size.
+ * Criteria: it shouldn't increase by more than threshold (@see ALLOWED_IMAGE_SIZE_INCREASE_THRESHOLD_PERCENT).
+ * @param currentName - name of original Docker image
+ * @param previousName - name of previous Docker image
+ * @return true if image size increase suppressed given threshold; false otherwise (including situation when ...
+ * ... it wasn't possible to determine any of image sizes)
+ */
+fun imageSizeIncreasedTooMuch(currentName: String, previousName: String): Boolean {
```
Comment (burnasheva) [inline tool/automation/ImageValidation.kts:159]: What about a decreasing image size? Should we also catch such situations?

Comment (AndreyKoltsov1997) [reply tool/automation/ImageValidation.kts:159]: Great idea - I think that'd be reasonable.
An unexpected removal or certain layer might indicate the change within one of our dependencies, for example: base image, utility tools, etc. Even if we wouldn't observe error during image creation and its bootstrap (entrypoint execution, start up of services), It **might** result into inability to run a particular script in runtime.
I will change the function to check percentage difference by modulo, not an absolute value.

**Comment by AndreyKoltsov1997** [comment]
Temporarily closing the PR to implement additional enhancements, such as more sufficient dependencies and handling of failures within steps.

**Pull request closed without merging**